modules/publisher/views.py

```diff
@@ -219,27 +219,31 @@ def publisher_snap_metrics(snap_name):
     except ApiError as api_error:
         return _handle_errors(api_error)
 
-    nodata = logic.has_data(metrics_response_json['metrics'])
+    active_devices = metrics.ActiveDevices(
+        name=metrics_response['metrics'][0]['metric_name'],
+        series=metrics_response['metrics'][0]['series'],
+        buckets=metrics_response['metrics'][0]['buckets'],
+        status=metrics_response['metrics'][0]['status'])
 
-    active_devices = {
-        'series': metrics_response_json['metrics'][0]['series'],
-        'buckets': metrics_response_json['metrics'][0]['buckets']
-    }
-    active_devices['series'] = sorted(
-        active_devices['series'],
-        key=itemgetter('name'))
-
-    latest_active_devices = logic.get_number_latest_active_devices(
-        active_devices)
+    latest_active = 0
+    if active_devices.has_data():
+        latest_active = active_devices.get_number_latest_active_devices()
 
-    users_by_country = public_logic.calculate_metrics_countries(
-        metrics_response_json['metrics'][1]['series'])
+    country_metric = metrics_helper.find_metric(
+        metrics_response['metrics'], "weekly_installed_base_by_country")
+    country_devices = metrics.CountryDevices(
+        name=country_metric['metric_name'],
+        series=country_metric['series'],
+        buckets=country_metric['buckets'],
+        status=country_metric['status'],
+        private=True)
 
-    country_data = public_logic.build_country_info(
-        users_by_country,
-        True)
+    territories_total = 0
+    if country_devices.has_data():
+        territories_total = country_devices.get_number_territories()
 
-    territories_total = logic.get_number_territories(country_data)
+    # TODO this is stupid
+    nodata = not (country_devices.has_data() or active_devices.has_data())
```

Comment:
If you wanted, `has_data` function could be used in `__bool__` which Python calls in truthy checks. Then an alternative to this is:
```
not any([country_devices, active_devices])
```

I'm not sure that solves your comment here but it is something fun.

Reply:
done